File: sol/solution/0001-0100/0010/Solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[0][0] = True
        for j in range(2, n + 1, 2):
            if p[j-1] == '*' and dp[0][j-2]:
                dp[0][j] = True
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[j-1] == '*':
                    dp[i][j] = dp[i][j-2] or (dp[i-1][j] and (s[i-1] == p[j-2] or p[j-2] == '.'))
                else:
                    dp[i][j] = dp[i-1][j-1] and (s[i-1] == p[j-1] or p[j-1] == '.')
        return dp[m][n]

File: sol/solution/0001-0100/0010/Solution.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        int m = s.Length, n = p.Length;
        bool[,] dp = new bool[m + 1, n + 1];
        dp[0, 0] = true;
        for (int j = 2; j <= n; j += 2) {
            if (p[j-1] == '*' && dp[0, j-2]) dp[0, j] = true;
        }
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j-1] == '*') {
                    dp[i, j] = dp[i, j-2] || (dp[i-1, j] && (s[i-1] == p[j-2] || p[j-2] == '.'));
                } else {
                    dp[i, j] = dp[i-1, j-1] && (s[i-1] == p[j-1] || p[j-1] == '.');
                }
            }
        }
        return dp[m, n];
    }
}